pkg/applicationserver/applicationserver.go

```diff
@@ -1171,94 +902,93 @@ func (as *ApplicationServer) handleDownlinkQueueInvalidated(ctx context.Context,
-	logger := log.FromContext(ctx)
-	pc, err := as.GetPeerConn(ctx, ttnpb.ClusterRole_NETWORK_SERVER, &ids)
-	if err != nil {
-		return err
-	}
-	client := ttnpb.NewAsNsClient(pc)
-	res, err := client.DownlinkQueueList(ctx, &ids, as.WithClusterAuth())
+	peer, err := as.GetPeer(ctx, ttnpb.ClusterRole_NETWORK_SERVER, &ids)
 	if err != nil {
-		logger.WithError(err).Warn("Failed to list the downlink queue for inserting nacked downlink message")
-		registerDropDownlink(ctx, ids, msg, err)
-	} else {
-		_, err := as.deviceRegistry.Set(ctx, ids,
-			[]string{
-				"pending_session",
-				"session",
-				"skip_payload_crypto_override",
-			},
-			func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
-				if dev == nil {
-					return nil, nil, errDeviceNotFound.WithAttributes("device_uid", unique.ID(ctx, ids))
-				}
-				mask, err := as.matchSession(ctx, ids, dev, link, msg.SessionKeyID)
-				if err != nil {
-					return nil, nil, err
-				}
-				if as.skipPayloadCrypto(ctx, link, dev, dev.Session) {
-					// When skipping application payload crypto, the upstream application is responsible for recalculating the
-					// downlink queue. No error is returned here to pass the downlink nack message upstream.
-					return dev, mask, nil
-				}
-				queue, _ := ttnpb.PartitionDownlinksBySessionKeyIDEquality(dev.Session.SessionKeyID, res.Downlinks...)
-				queue = append([]*ttnpb.ApplicationDownlink{msg}, queue...)
-				if err := as.recalculateDownlinkQueue(ctx, dev, link, dev.Session, queue, msg.FCnt+1, false); err != nil {
-					return nil, nil, err
-				}
-				return dev, append(mask, "session"), nil
-			},
-		)
-		if err != nil {
-			logger.WithError(err).Warn("Failed to recalculate downlink queue with inserted nacked downlink message")
-			registerDropDownlink(ctx, ids, msg, err)
-		}
-	}
-	// Decrypt the message as it will be sent to upstream after handling it.
-	if err := as.decryptDownlinkMessage(ctx, ids, msg, link); err != nil {
 		return err
 	}
-	return nil
+	_, err = as.deviceRegistry.Set(ctx, ids,
+		[]string{
+			"pending_session",
+			"session",
+			"skip_payload_crypto_override",
+		},
+		func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+			if dev == nil {
+				return nil, nil, errDeviceNotFound.WithAttributes("device_uid", unique.ID(ctx, ids))
+			}
+
+			if as.skipPayloadCrypto(ctx, link, dev, dev.Session) {
+				// When skipping application payload crypto, the upstream application is responsible for recalculating the
+				// downlink queue. No error is returned here to pass the downlink nack message upstream.
+				return dev, nil, nil
+			}
+
+			matchMask, err := as.matchSession(ctx, ids, dev, link, msg.SessionKeyID)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			// Decrypt the message as it will be sent to upstream after handling it.
+			if err := as.decryptDownlinkMessage(ctx, ids, msg, link); err != nil {
+				return nil, nil, err
+			}
+
+			items := []*ttnpb.ApplicationDownlink{msg}
```

Comment:
Should we already bump the FCnt here?

Reply:
`attemptDownlinkQueueOp` will bump it automatically.